src/CatalogApplication/Controllers/TagController.cs

```diff
@@ -1,130 +1,174 @@
+            FeedIterator<Tag> iterator = _dbService.tagContainer.GetItemQueryIterator<Tag>(queryDef);
+
+            List<Tag> tags = new List<Tag>();
+
+            while (iterator.HasMoreResults)
+            {
+                var resultSet = await iterator.ReadNextAsync();
+                foreach (Tag tag in resultSet)
+                {
+                    tags.Add(tag);
+                }
+            }
+
+            return tags;
+        }
+
+
+        /// <summary>
+        /// Retrieve a list of tags by noteId.
+        /// </summary>
+        /// <param name="noteId">The note ID</param>
+        /// <returns>An array containing a value determined by the parameter</returns>
+        [HttpGet("{noteId}")]
+        public async Task<List<Tag>> GetTagsByNote(string noteId)
+        {
+            return await GetTags($"SELECT * FROM c WHERE c.noteId={noteId}");
+        }
+
+        /// <summary>
+        /// Gets tags by noteId and userId
+        /// </summary>
+        /// <param name="noteId">The noteId to look for the tag</param>
+        /// <param name="userId">The userId to look for the tag</param>
+        /// <returns>A boolean indicating whether a tag is created or not</returns>
+        [HttpGet("{noteId}/users/{userId}")]
+        public async Task<List<Tag>> GetTagsByNoteAndUser(string noteId, string userId)
+        {
+            return await GetTags($"SELECT * FROM c WHERE c.noteId={noteId} AND c.userId={userId}");
+        }
+
+        /// <summary>
+        /// Create a new tag. Pass everything other than id.
+        /// </summary>
+        /// <returns>OK if successful otherwise bad request</returns>
+        [HttpPost]
+        public async Task<IActionResult> CreateNewTag(Tag tag)
+        {
+            try
+            {
+                if (ModelState.IsValid)
+                {
+                    tag.id = System.Guid.NewGuid().ToString();
+                    await _dbService.tagContainer.CreateItemAsync(tag);
+                }
+                return Ok();
+            }
+            catch
+            {
+                return BadRequest();
+            }
+        }
+
+        /// <summary>
+        /// Updates a tag based off of id. 
+        /// </summary>
+        /// <param name="id">The id to look for the specific tag</param>
+        /// <param name="tag">the updated version of the tag</param>
+        /// <returns></returns>
+        [HttpPut("{id}")]
+        [ActionName("Edit")]
+        public async Task<ActionResult> UpdateTag(string id, Tag tag)
+        {
+            if (id == null)
+            {
+                return BadRequest();
+            }
+            if (ModelState.IsValid)
+            {
+                try
+                {
+                    await _dbService.tagContainer.UpsertItemAsync<Tag>(tag, new PartitionKey(id));
+                    return Ok();
+                }
+                catch
+                {
+                    return BadRequest();
```

Comment:
This isn't necessarily a bad request outcome. What if the database service is down?

In general it's a good idea to create a response wrapper which handles errors and other issues after API calls are processed. The response wrapper hides information which callers shouldn't have, such as details of how the internals of your service works, and exposes relevant information for user action. This is something you all can consider later though.